Exception-Handling.md

```diff
@@ -0,0 +1,254 @@
+
+You can see the results in the error message:
+
+```bash
+multimethod.rb:2:in `/': divided by 0 (ZeroDivisionError)
+	from multimethod.rb:2:in `quotient_and_remainer'
+	from multimethod.rb:11:in `block in mystery'
+	from multimethod.rb:10:in `each'
+	from multimethod.rb:10:in `mystery'
+	from multimethod.rb:17:in `<main>'
+```
+
+As you can see the error messages display in order as Ruby falls back from where it encountered the error, moving, "through the stack," unwinding the method calls that lead to the error.  
+
+
+
+#### Exception Handling in Rails
+
+In Rails we have a number of methods in ActiveRecord which can generate exceptions a few of which include:
+-  save!
+-  create!
+-  update!
+
+These methods trigger exceptions when validations fail or the database is unavailable.  So you could handle problems with creating a new task in TaskListRails using exceptions as follows.  
+
+```ruby
+  def create
+    @task = Task.new(task_params)
+
+    begin
+      @task.save!
+    rescue ActiveRecord::RecordInvalid
+      flash[:notice] = "Cannot Save the Task."
+    ensure
+      redirect_to tasks_new_path
+    end
+  end
+```
+
+However while this can be done it isn't good practice.  It's considered best practice to [**not use exceptions for expected outcomes**](https://robots.thoughtbot.com/save-bang-your-head-active-record-will-drive-you-mad).  Since we expect users to give us invalid input on a regular basis, the above example does not qualify.  This leaves us with the question:  
+
+### Why Would We Use Rescue?
+
+There are a lot of [differing opinions](https://www.sitepoint.com/ruby-error-handling-beyond-basics/) on when and why you should implement exception handling in your code.  The consensus seems to be that exception handling should be for *exceptional* circumstances.  In other words you use Exception Handling (rescue) for abnormal events, and regular if statements for normal occurrences like validation errors.
+
+You can think of your program errors falling into two categories:
+-  Things you expect to happen regularly, like:
+	- Users putting invalid items into forms.
+	- Users trying to access things they shouldn't have access to.
+-  Errors that happen very rarely like:
+	- Database connection errors
+	- API failures
+
+##### So you would use If Statements for:
+- Invalid user input (validation errors on Active Records Etc).  
+- Users trying to access resources they do not have permission for.
+
+##### You would use Exception Handling for:
+-  Database failures
+-  API service Failures
+	- For example when Github is unavailable in a service it provides.  
+	
+So you could rewrite the Task Controller's create method as:
+```ruby
+ def create
+    @task = Task.new(task_params)
+
+    begin
+      if ! @task.save
+        msg = ""
+        @task.errors.each do |field, message|
+          msg += "#{field.to_s.capitalize}: #{message}"
+        end
+        flash[:notice] = msg
```

Comment:
I know, we can at least talk about the logger, but I'm also afraid about this getting too complicated already.